src/machines.jl

```diff
@@ -78,39 +79,192 @@ constructor for retraining on new data using the saved model.
+
+    # instantiate node and machine dictionaries:
+    newnode_given_old =
+        IdDict{AbstractNode,AbstractNode}([old => source() for old in sources(W)])
+    
+    newoperation_node_given_old =
+        IdDict{AbstractNode,AbstractNode}()
+    newreport_node_given_old =
+        IdDict{AbstractNode,AbstractNode}()
+    newmach_given_old = IdDict{Machine,Machine}()
+
+    # build the new network, nodes are nicely ordered
+    for N in nodes_
+        # Retrieve the future node's ancestors
+        args = [newnode_given_old[arg] for arg in N.args]
+        if N.machine === nothing
+            newnode_given_old[N] = node(N.operation, args...)
+        else
+            # The same machine can be associated with multiple nodes
+            if N.machine in keys(newmach_given_old)
+                m = newmach_given_old[N.machine]
+            else
+                m = serializable(filename, N.machine)
+                m.args = Tuple(newnode_given_old[s] for s in N.machine.args)
+                newmach_given_old[N.machine] = m
+            end
+            newnode_given_old[N] = N.operation(m, args...)
+        end
+        # Sort nodes according to: operation_node, report_node
+        if N in operation_nodes
+            newoperation_node_given_old[N] = newnode_given_old[N]
+        elseif N in report_nodes
+            newreport_node_given_old[N] = newnode_given_old[N]
+        end
     end
-    mach.state = 1
-    mach.fitresult = fitresult
-    mach.report = report
+
+    newoperation_nodes = Tuple(newoperation_node_given_old[N] for N in
+            operation_nodes)
+    newreport_nodes = Tuple(newreport_node_given_old[N] for N in
+            report_nodes)
+    report_tuple =
+        NamedTuple{keys(MLJBase._report_part(signature))}(newreport_nodes)
+    operation_tuple =
+        NamedTuple{keys(MLJBase._operation_part(signature))}(newoperation_nodes)
+
+    newsignature = if isempty(report_tuple)
+                        operation_tuple
+                    else
+                        merge(operation_tuple, (report=report_tuple,))
+                    end
+    
+
+    newfitresult = MLJBase.CompositeFitresult(newsignature)
+    setfield!(newfitresult, :report_additions, report_tuple)
+
+    return newfitresult
+
+end
+
+###############################################################################
+#####                  SERIALIZABLE AND RESTORE                           #####
+###############################################################################
+
+
+"""
+    serializable(filename, mach::Machine; kwargs...)
+
+Copies the state of the machine to make it serializable.
+    - Removes all data from caches dans data fields
+    - Makes all `fitresults` serializable
+"""
+function serializable(filename, mach::Machine; kwargs...)
+    copymach = machine(mach.model, mach.args..., cache=typeof(mach).parameters[2])
+
+    for fieldname in fieldnames(Machine)
+        if fieldname ∈ (:model, :args, :report)
+            continue
```

Comment:
Shouldn't we also do `copymach.args = ()` to empty the training arguments? For example, in an "ordinary" (non-composite) machine, these arguments are the training data wrapped in source nodes.

Reply:
Good catch, I thought the source nodes were actually pulling from the `data` field of the machine.